presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,397 @@
+                    // equal to min representable value
+                    switch (operator) {
+                        case LESS_THAN:
+                            return falseIfNotNull(cast.getExpression());
+                        case LESS_THAN_OR_EQUAL:
+                            return new ComparisonExpression(EQUAL, cast.getExpression(), literalEncoder.toExpression(min, sourceType));
+                        case GREATER_THAN_OR_EQUAL:
+                            return trueIfNotNull(cast.getExpression());
+                        case GREATER_THAN:
+                            return new ComparisonExpression(NOT_EQUAL, cast.getExpression(), literalEncoder.toExpression(min, sourceType));
+                        case EQUAL:
+                        case NOT_EQUAL:
+                        case IS_DISTINCT_FROM:
+                            return new ComparisonExpression(operator, cast.getExpression(), literalEncoder.toExpression(min, sourceType));
+                        default:
+                            throw new UnsupportedOperationException("Not yet implemented: " + operator);
+                    }
+                }
+            }
+
+            Signature targetToSource;
+            try {
+                targetToSource = metadata.getFunctionRegistry().getCoercion(targetType, sourceType);
+            }
+            catch (OperatorNotFoundException e) {
+                // Without a cast between target -> source, there's nothing we can do
+                return expression;
+            }
+
+            Object literalInSourceType;
+            try {
+                literalInSourceType = coerce(right, targetToSource);
+            }
+            catch (PrestoException e) {
+                // A failure to cast from target -> source type could be because:
+                //  1. missing cast
+                //  2. bad implementation
+                //  3. out of range or otherwise unrepresentable value
+                // Since we can't distinguish between those cases, take the conservative option
+                // and bail out.
+                return expression;
+            }
+
+            Object roundtripLiteral = coerce(literalInSourceType, sourceToTarget);
+
+            int literalVsRoundtripped = compare(targetType, right, roundtripLiteral);
+
+            if (literalVsRoundtripped > 0) {
+                // cast rounded down
+                switch (operator) {
+                    case EQUAL:
+                        return falseIfNotNull(cast.getExpression());
+                    case NOT_EQUAL:
+                        return trueIfNotNull(cast.getExpression());
+                    case IS_DISTINCT_FROM:
+                        return TRUE_LITERAL;
+                    case LESS_THAN:
+                    case LESS_THAN_OR_EQUAL:
+                        if (sourceType.getRange().isPresent() && compare(sourceType, sourceType.getRange().get().getMin(), literalInSourceType) == 0) {
```

Comment:
nit, reuse variable:
```suggestion
                        if (range.isPresent() && compare(sourceType, range.get().getMin(), literalInSourceType) == 0) {
```
(+ consider renaming the variable `range → sourceRange`)